### PR TITLE
C# 12: Primary constructor inititalizers.

### PR DESCRIPTION
**Pull request by michaelnebel**
It turns out that the extraction of primary constructors didn't extract initialisers correctly. This is fixed in this PR.

That is, for
```csharp
public class C1(object o1) { }
public class C2(object o2) : C1(o2) { }
```
We need to extract
- The *implicit* inititaliser (calling the `object` constructor) for the primary constructor on `C1`.
- The *explicit* initialiser (calling the `C1` constructor) for the primary constructor on `C2`.

This is needed for
- Getting data-flow to work properly for *simple* classes with primary constructors (otherwise we won't get a control flow node corresponding to the basic block containing the class declaration). Adding the implicit initialiser means that we at least have a first statement, which is calling the `object` constructor.
- First step for getting data flow to work for classes with primary constructors that uses inheritance (the explicit initialiser needs to be extracted otherwise we can't detect flow from the primary constructor of `C2` into the `C1` base class constructor).

It also turns out that we also need to extract at least an empty (synthetic) body for primary constructors to avoid lots of control flow issues.

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs [hunk 1]
```diff
@@ -10,8 +12,16 @@ namespace Semmle.Extraction.CSharp.Entities
 {
     internal class Constructor : Method
     {
+        private readonly Lazy<List<SyntaxNode>> DeclaringReferenceSyntax;
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs:15]: What do we gain with `Lazy<>`? I think we're using these nodes in the `ReportingLocation`. Doesn't this mean that we need the `DeclaringReferenceSyntax` for all constructors?

Minor: fields should start with lower case letters.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs [hunk 2]
```diff
@@ -22,6 +32,12 @@ public override void Populate(TextWriter trapFile)
             trapFile.constructors(this, Symbol.ContainingType.Name, ContainingType, (Constructor)OriginalDefinition);
             trapFile.constructor_location(this, Location);
 
+            if (IsPrimary)
+            {
+                // Create a synthetic empty body for primary constructors.
+                Statements.SyntheticEmptyBlock.Create(Context, this, 0, Location);
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs:38]: As discussed on Slack, this might also be useful for default constructors.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs [hunk 3]
```diff
@@ -33,68 +49,79 @@ public override void Populate(TextWriter trapFile)
         protected override void ExtractInitializers(TextWriter trapFile)
         {
             // Do not extract initializers for constructed types.
-            if (!IsSourceDeclaration)
+            // Only extract initializers for constructors with a body and primary constructors.
+            if (Block is null && ExpressionBody is null && !IsPrimary ||
+                !IsSourceDeclaration)
+            {
                 return;
+            }
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs:57]: As discussed on Slack, this might also apply to default constructors too.
```
class Base {}
class Derived : Base {}
```
`Derived` has a default constructor, which implicitly is calling `base()`.

File: csharp/ql/test/library-tests/constructors/PrintAst.expected [hunk 4]
```diff
@@ -25,6 +25,7 @@ constructors.cs:
 #   23|           -1: [TypeMention] object
 #   23|         1: [Parameter] s
 #   23|           -1: [TypeMention] string
+#   23|       4: [BlockStmt] {...}
```
Comment (tamasvajk) [inline csharp/ql/test/library-tests/constructors/PrintAst.expected:28]: For generated expressions, we're using `expr_compiler_generated`. Should we mark `SyntheticEmptyBlock`s as compiler generated too?

**Review by tamasvajk — COMMENTED** [review]
I've added some minor comments/question.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs (hunk 1 above)
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs:15]: Good point, I will just initialize it directly in the constructor.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs (hunk 2 above)
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs:38]: Yes, that is a good point!
I would prefer to make furthermore improvements on follow up PRs. Is that ok?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs (hunk 3 above)
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs:57]: same comment as above (and thank you for putting it here!)

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs (hunk 2 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Constructor.cs:38]: Yep, a separate PR sounds good.

File: csharp/ql/test/library-tests/constructors/PrintAst.expected (hunk 4 above)
Comment (michaelnebel) [reply csharp/ql/test/library-tests/constructors/PrintAst.expected:28]: We can add it to the `compiler_generated` table?

Comment (michaelnebel) [reply csharp/ql/test/library-tests/constructors/PrintAst.expected:28]: Hmm, that will require database changes.

Comment (michaelnebel) [reply csharp/ql/test/library-tests/constructors/PrintAst.expected:28]: We could create `stmt_compiler_generated`. Should we do that?

Comment (tamasvajk) [reply csharp/ql/test/library-tests/constructors/PrintAst.expected:28]: I think it would be fine to do this in a separate PR, so that you could merge this one. 

Is there any benefit of having `expr_compiler_generated`, `compiler_generated`, and `stmt_compiler_generated`? It looks like any solution would need a DB schema change. I think my preference would be merging all of this into `compiler_generated`.

Comment (michaelnebel) [reply csharp/ql/test/library-tests/constructors/PrintAst.expected:28]: That would be a nice simplification in any case. I will add that to my list of follow up items!

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 5]
```diff
@@ -953,12 +953,8 @@ private module Cached {
...
     } or
     TFlowInsensitiveFieldNode(FieldOrProperty f) { f.isFieldLike() } or
-    TInstanceParameterAccessNode(ControlFlow::Node cfn, boolean isPostUpdate) {
-      exists(ParameterAccess pa | cfn = getAPrimaryConstructorParameterCfn(pa) |
-        isPostUpdate = false
-        or
-        pa instanceof ParameterWrite and isPostUpdate = true
-      )
+    TInstanceParameterAccessNode(ControlFlow::Node cfn, Boolean isPostUpdate) {
+      cfn = getAPrimaryConstructorParameterCfn(_)
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:957]: Why was this changed? It seems that we are now creating redundant post-update nodes for parameter reads.

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:957]: Without post-update nodes this lead to a failing dataflow consistency test for (according to the `reverseRead`)
```csharp
class ClassC2(object oc2) : ClassC1(oc2) { }
```
Should we add these as exceptions?

**Comment by michaelnebel** [comment]
I realized that I had forgotten to start the DCA experiment with flags to rebuild the extractor.
The above experiment is based on the merge commit and the parent from main.
It shows some new results, but only in quality queries, mostly in `useless-type-test`, but they seem unrelated to the changes (but they look sound) and they are all in generated code.
